File: haskell/P2/P2.hs
module P2 (p2, fib) where

fib :: [Int]
fib = 1 : 2 : zipWith (+) fib (tail fib)

p2 :: Int
p2 = sum $ takeWhile (<4000000) [ x | x <- fib, even x ]

main = print p2

File: scala/projecteuler/src/main/scala/P2.scala
package com.example.projecteuler

class P2 {
  lazy val fib: Stream[Int] = Stream.cons(1, Stream.cons(2, fib.zip(fib.tail).map(p => p._1 + p._2)))
  def solve() : Int = {
    fib.filter(_ % 2 == 0).takeWhile(_ < 4000000).sum
  }
}
